### ISSUE TITLE
Converting between unions of TypedDicts with unpacking

### ISSUE THREAD

**Issue by chris-leach** [microsoft/pyright#7677]
**Is your feature request related to a problem? Please describe.**
I'm having problems constructing `TypedDict`s by unpacking, where there seems to be sufficient information to prove compliance.

Consider the following:

```python
from datetime import UTC, datetime
from typing import Literal, TypedDict


class StrContentDict(TypedDict):
    content_type: Literal["str"]
    content: str


class IntContentDict(TypedDict):
    content_type: Literal["int"]
    content: int


ContentDict = StrContentDict | IntContentDict


class StrJobDict(TypedDict):
    timestamp: datetime
    content_type: Literal["str"]
    content: str


class IntJobDict(TypedDict):
    timestamp: datetime
    content_type: Literal["str"]
    content: str


JobDict = StrJobDict | IntJobDict


def job_from_content(content: ContentDict) -> JobDict:
    return {  # error! reportReturnType
        **content,
        "timestamp": datetime.now(UTC),
    }
```
This appears to be correct, since `content_type` and `content` will necessarily be the correct types for both variants of input and output, but it results in the error:
```
Expression of type "dict[str, object | datetime]" cannot be assigned to return type "JobDict"
  Type "dict[str, object | datetime]" cannot be assigned to type "JobDict"
    "dict[str, object | datetime]" is incompatible with "StrJobDict"
    "dict[str, object | datetime]" is incompatible with "IntJobDict"
```

**Describe the solution you’d like**
Ideally pyright would detect that the dict comprehension cannot result in a non-compliant output given compliant function input.

More generally, following on from the discussion in #7574, am I going against the grain of the typing spec by trying to use `TypedDict` for structural typing in this way? I guess I'm comparing to TypeScript, which has an actual structural type system and where something like the following works:

<details>
<summary>TypeScript code</summary>

```typescript
type StrContent = {
  contentType: "str";
  content: string;
};

type IntContent = {
  contentType: "int";
  content: number;
};

type Content = StrContent | IntContent;

type StrJob = {
  timestamp: Date;
  contentType: "str";
  content: string;
};

type IntJob = {
  timestamp: Date;
  contentType: "int";
  content: number;
};

type Job = StrJob | IntJob;

function jobFromContent(content: Content): Job {
  return {
    ...content,
    timestamp: new Date(),
  };
}
```

</details>

 It seems like the whole purpose of `TypedDict` is to allow at least a degree of structural typing, but there are quite a few rough edges.

**Comment by erictraut** [comment]
What you're proposing here would require some very ugly special-casing logic and would narrowly work only in very specific cases. I don't plan to implement this unless there's strong signal from pyright users that this _very specific_ use case is _very common_. I think that's unlikely to happen.

Not surprisingly, mypy and the other major Python type checkers also generate a type error here.

Even in typescript, this appears to be special-cased in a pretty narrow way. For example, consider the small tweak to your typescript code sample. This doesn't type check without errors.

```ts
function jobFromContent(content: ['str', string] | ['int', number]): Job {
    return {
        contentType: content[0],
        content: content[1],
        timestamp: content,
    };
```

It appears that the TS compiler has special-cased the specific situation where an object literal expression makes use of an unpack operator applied to a simple variable whose type is a union of two or more structural types. In that case, it evaluates the resulting structural type as a union of multiple types rather than combining them into a single (non-unioned) structural type. If you stray a little ways from that pattern, as I've done above, it collapses back to a single type rather than a union.

This is also complicated by differences between the structural types in TypeScript and the `TypedDict` found in the Python type system. TypeScript is a bit "looser" in its rules. This makes TypeScript less cumbersome to use, but it allows some type errors to go undetected. The folks who designed `TypedDict` opted for a stricter set of rules. These make `TypedDict` a bit harder to use but more airtight from a type checking perspective. (FWIW, I prefer the TypeScript approach.) Here's an example of some code in TypeScript that contains an undetected type error:

```ts
function jobFromContent(content: Content): Job {
  return {
    timestamp: new Date(),
    ...content,
  };
}

function makeStrContent(content: string, otherFields: object): StrContent {
    return {
        contentType: 'str',
        content,
        ...otherFields,
    };
}

const job = jobFromContent(makeStrContent('hello', { timestamp: 'not a date' }));
console.log(job.timestamp) // This is not a Date as you would expect!
```

There is a draft [PEP 728](https://peps.python.org/pep-0728/) that adds the concept of a "closed" `TypedDict`. This will allow us to do some things with `TypedDict` that are less cumbersome but still type safe. Pyright has provisional (experimental) support for PEP 728 if you'd like to play with it. The PEP hasn't yet been submitted for formal approval by the typing council or the steering council.

I'm going to close this enhancement request for now because I think it's too much of a special case and will not benefit enough pyright users to justify the complexity involved. I'm willing to revisit this in the future if there's evidence to the contrary.

**Comment by chris-leach** [comment]
Thanks for the detailed explanation.
